File: 0713-subarray-product-less-than-k/solution.java
// class Solution {
//     public int numSubarrayProductLessThanK(int[] nums, int k) {
//         if(k == 0) return 0;
//         int cnt=0;
//         for(int i=0; i<nums.length; i++){
//             long prod=1;
//             for(int j=i; j<nums.length; j++){
//                 prod*=nums[j];
//                 if(prod < k)
//                   cnt++;
//                 else break;
//             }
//         }
//         return cnt;
//     }
// }

class Solution {
    public int numSubarrayProductLessThanK(int[] nums, int k) {
        // @ yt- code bashers
        if(k == 0) return 0;
        int cnt=0, end = 0, start=0, product = 1;
        while(end < nums.length){
            product*=nums[end];

            while(start < end && product >= k){
                product/=nums[start];
                start++;
            }

            if( product < k) cnt+=end-start+1;

            end++;
        }
        return cnt;
    }
}

File: 0713-subarray-product-less-than-k/solution.cpp
class Solution {
public:
    int numSubarrayProductLessThanK(vector<int>& nums, int k) {
        // vvvi base case // a/c to constraits & problem statement
        if(k <= 1) return 0;  // vvvi to handle 


        int start = 0, productSoFar = 1, cnt = 0;

        for(int end = 0; end<nums.size(); end++){
            productSoFar*=nums[end];
            while(productSoFar >= k){
                productSoFar/=nums[start];
                start++;
            }
            cnt+=(end-start+1);
        }
        return cnt;
    }
};
